src/util.py

```diff
@@ -0,0 +1,320 @@
+        all_fields.update(s.fields)
+
+    # Group by name to find conflicts
+    fields_by_name = collections.defaultdict(list)
+    for field in all_fields:
+        fields_by_name[field.name].append(field)
+
+    err_count = 0
+    for f_name, fields in fields_by_name.items():
+        if len(fields) == 1:
+            continue
+
+        type_set = {f.dataType for f in fields}
+        # Cast long ints to doubles
+        if type_set == {LongType(), DoubleType()}:
+            diff_fields.append(StructField(f_name, DoubleType()))
+        # Cast ints to booleans
+        elif type_set == {LongType(), BooleanType()}:
+            diff_fields.append(StructField(f_name, DoubleType()))
+        else:
+            err_count += 1
+            print(f"{f_name} has multiple types that can't be resolved:")
+            print(f"  {type_set}")
+
+    if err_count:
+        raise ValueError(f"Could not resolve {err_count} field(s)")
+
+    # Return just the fields that might have conflicts
+    return StructType(diff_fields)
+
+
+def load_clean_and_add_upgrade_id(
+    file_path: str,
+    upgrade_id: float,
+    schema: StructType = None,
+    schema_diffs: StructType = None,
+):
+    """
+    Load a DataFrame from a Parquet file, clean its column names, and add an upgrade_id column.
+
+    Args:
+    - file_path (str): Path to the Parquet file
+    - upgrade_id (double): The value for the 'upgrade_id' column
+    - schema (StructType, optional): Optional schema to enforce on the DataFrame
+    - schema_diffs (StructType, optional): Schema containing fields that should be cast to other
+        types if they don't already match.
+
+    Returns:
+    - DataFrame with cleaned column names and upgrade_id column if provided
+    """
+
+    if schema is not None:
+        df = spark.read.schema(schema).parquet(str(file_path))
+    else:
+        df = spark.read.parquet(str(file_path))
+
+    def get_dtype(df, colname):
+        """Get the type of a column within a dataframe"""
+        return [dtype for name, dtype in df.dtypes if name == colname][0]
+
+    for field in schema_diffs.fields:
+        if field.name in df.schema.fieldNames():
+            if field.dataType.typeName() != get_dtype(df, field.name):
+                # Escape column names with periods, because PySpark doesn't like them.
+                # These are cleaned up later.
+                colname = f"`{field.name}`" if "." in field.name else field.name
+                df = df.withColumn(field.name, df[colname].cast(field.dataType))
+
+    if "completed_status" in df.columns:
+        df = df.where(F.col("completed_status") == "Success")
+
+    df = df.withColumn("upgrade_id", F.lit(upgrade_id))
+
+    return df
+
+
+def read_combine_sims():
+    """
+    Read in RAStock sims and combine into a single Spark DF
+
+    This function is getting revamped in a different repo so not bothering to document for now
+    """
+    # TODO: update with new code from natalie/mohammad in pep
```

Comment:
This changed already in https://github.com/rewiringamerica/pep/pull/187, but if you want to wait for it to finish changing, this is fine for now.

Reply:
yes, i saw that but I thought I'd wait till the upgrade id table was ready and then make all the changes in one go.